pkg/loadbalance/interleaved_weighted_round_robin.go

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2023 CloudWeGo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package loadbalance
+
+import (
+	"context"
+	"sync"
+
+	"github.com/bytedance/gopkg/lang/fastrand"
+
+	"github.com/cloudwego/kitex/pkg/discovery"
+)
+
+type iwrrNode struct {
+	discovery.Instance
+	remainder int
+
+	next *iwrrNode
+}
+
+type iwrrQueue struct {
+	head *iwrrNode
+	tail *iwrrNode
+}
+
+type InterleavedWeightedRoundRobinPicker struct {
+	current *iwrrQueue
+	next    *iwrrQueue
+	gcd     int
+
+	lock sync.Mutex
+}
+
+func newInterleavedWeightedRoundRobinPicker(instances []discovery.Instance) Picker {
+	iwrrp := new(InterleavedWeightedRoundRobinPicker)
+	iwrrp.current = newIwrrQueue()
+	iwrrp.next = newIwrrQueue()
+
+	size := uint64(len(instances))
+	offset := fastrand.Uint64n(size)
+	gcd := 0
+	for idx := uint64(0); idx < size; idx++ {
+		ins := instances[(idx+offset)%size]
+		gcd = gcdInt(gcd, ins.Weight())
+
+		iwrrp.current.enqueue(&iwrrNode{
+			Instance:  ins,
+			remainder: ins.Weight(),
+		})
+	}
+
+	iwrrp.gcd = gcd
+
+	return iwrrp
+}
+
+func (ip *InterleavedWeightedRoundRobinPicker) Next(ctx context.Context, request interface{}) discovery.Instance {
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
+
+	if ip.current.empty() {
+		ip.current, ip.next = ip.next, ip.current
+	}
+
+	node := ip.current.dequeue()
+	node.remainder -= ip.gcd
+
+	if node.remainder > 0 {
+		ip.current.enqueue(node)
```

Comment:
A random switch with another instance in current will make this algorithm much smoother I guess

Reply:
I always think that smoothness is not so important in high-traffic scenarios, because as above, in high-traffic scenarios, a cycle of selection will be completed in short time. At this time, no matter whether it is smooth or not, there is no obvious perception of service quality. As for the small traffic scene, who cares if it is smooth?